File: src/comms/mqtt/MQTTBroker.cpp
#include "comms/mqtt/MQTTBroker.hpp"
#include <algorithm>

XBT_LOG_NEW_DEFAULT_CATEGORY(mqtt_broker, "MQTT Broker");

namespace enigma {
namespace mqtt {

namespace sg4 = simgrid::s4u;

MQTTBroker::MQTTBroker(const std::string& name)
    : broker_name(name), messages_published(0), messages_delivered(0), running(true) {
    control_mbox = sg4::Mailbox::by_name(get_broker_mailbox(name));
}

void MQTTBroker::operator()() {
    sg4::Host* host = sg4::this_actor::get_host();
    XBT_INFO("MQTT Broker '%s' started on host '%s'", 
             broker_name.c_str(), host->get_cname());
    
    while (running) {
        // Receive control message
        auto* ctrl_msg = control_mbox->get<MQTTControlMessage>();
        
        switch (ctrl_msg->type) {
            case MQTTControlMessage::Type::SUBSCRIBE:
                handle_subscribe(ctrl_msg->topic, ctrl_msg->subscriber);
                break;
                
            case MQTTControlMessage::Type::UNSUBSCRIBE:
                handle_unsubscribe(ctrl_msg->topic, ctrl_msg->subscriber);
                break;
                
            case MQTTControlMessage::Type::PUBLISH:
                handle_publish(ctrl_msg->message);
                break;
                
            case MQTTControlMessage::Type::SHUTDOWN:
                XBT_INFO("Broker shutdown requested");
                running = false;
                break;
                
            default:
                XBT_WARN("Unknown control message type");
                break;
        }
        
        delete ctrl_msg;
    }
    
    print_stats();
    XBT_INFO("MQTT Broker '%s' terminated", broker_name.c_str());
}

void MQTTBroker::handle_subscribe(const std::string& topic, const std::string& subscriber) {
    auto& subs = subscriptions[topic];
    
    // Check if already subscribed
    if (std::find(subs.begin(), subs.end(), subscriber) == subs.end()) {
        subs.push_back(subscriber);
        XBT_INFO("Subscriber '%s' subscribed to topic '%s' (%zu total subscribers)",
                 subscriber.c_str(), topic.c_str(), subs.size());
    } else {
        XBT_DEBUG("Subscriber '%s' already subscribed to topic '%s'",
                  subscriber.c_str(), topic.c_str());
    }
}

void MQTTBroker::handle_unsubscribe(const std::string& topic, const std::string& subscriber) {
    auto it = subscriptions.find(topic);
    if (it != subscriptions.end()) {
        auto& subs = it->second;
        auto sub_it = std::find(subs.begin(), subs.end(), subscriber);
        
        if (sub_it != subs.end()) {
            subs.erase(sub_it);
            XBT_INFO("Subscriber '%s' unsubscribed from topic '%s'",
                     subscriber.c_str(), topic.c_str());
            
            // Remove topic if no more subscribers
            if (subs.empty()) {
                subscriptions.erase(it);
                XBT_DEBUG("Topic '%s' removed (no subscribers)", topic.c_str());
            }
        }
    }
}

void MQTTBroker::handle_publish(std::shared_ptr<MQTTMessage> msg) {
    messages_published++;
    
    XBT_INFO("Publishing message to topic '%s' (size: %zu bytes, from: %s)",
             msg->topic.c_str(), msg->size, msg->publisher.c_str());
    
    // Find subscribers for this topic
    auto it = subscriptions.find(msg->topic);
    if (it == subscriptions.end() || it->second.empty()) {
        XBT_DEBUG("No subscribers for topic '%s'", msg->topic.c_str());
        return;
    }
    
    // Deliver to all subscribers
    for (const auto& subscriber : it->second) {
        auto* mbox = sg4::Mailbox::by_name(subscriber);
        
        // Create a copy of the message for each subscriber
        auto* msg_copy = new std::shared_ptr<MQTTMessage>(msg);
        
        XBT_DEBUG("Delivering message to subscriber '%s'", subscriber.c_str());
        mbox->put(msg_copy, msg->size);
        messages_delivered++;
    }
    
    XBT_INFO("Message delivered to %zu subscribers", it->second.size());
}

std::string MQTTBroker::get_broker_mailbox(const std::string& broker_name) {
    return "mqtt_broker_" + broker_name;
}

std::string MQTTBroker::get_topic_mailbox(const std::string& broker_name,
                                          const std::string& topic) {
    return "mqtt_" + broker_name + "_topic_" + topic;
}

void MQTTBroker::print_stats() const {
    XBT_INFO("=== MQTT Broker Statistics ===");
    XBT_INFO("  Messages published: %d", messages_published);
    XBT_INFO("  Messages delivered: %d", messages_delivered);
    XBT_INFO("  Active topics: %zu", subscriptions.size());
    
    for (const auto& [topic, subs] : subscriptions) {
        XBT_INFO("    Topic '%s': %zu subscribers", topic.c_str(), subs.size());
    }
}

} // namespace mqtt
} // namespace enigma

File: src/python/enigma/comms/mqtt/mqtt_broker.py
"""
MQTTBroker: SimGrid-based MQTT broker simulation.

Runs as a SimGrid actor on a designated host.
Uses SimGrid Mailboxes to simulate the pub/sub message routing.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, List, Optional
import simgrid


# --------------------------------------------------------------------------- #
# Data structures                                                              #
# --------------------------------------------------------------------------- #

@dataclass
class MQTTMessage:
    """In-simulation MQTT message."""
    topic: str
    payload: str
    size: int                       # bytes
    publisher: str
    qos: int = 0
    timestamp: float = 0.0

    def __post_init__(self) -> None:
        if self.timestamp == 0.0:
            try:
                self.timestamp = simgrid.Engine.clock
            except Exception:
                pass


class _ControlType:
    SUBSCRIBE   = "SUBSCRIBE"
    UNSUBSCRIBE = "UNSUBSCRIBE"
    PUBLISH     = "PUBLISH"
    SHUTDOWN    = "SHUTDOWN"


@dataclass
class _ControlMessage:
    ctrl_type: str
    topic: str = ""
    subscriber: str = ""
    message: Optional[MQTTMessage] = None


# --------------------------------------------------------------------------- #
# MQTTBroker actor                                                             #
# --------------------------------------------------------------------------- #

class MQTTBroker:
    """
    Central MQTT broker.  Run as a SimGrid actor via :func:`start_broker`.

    The broker listens on a well-known mailbox
    ``<broker_name>_ctrl`` for :class:`_ControlMessage` objects and
    routes :class:`MQTTMessage` objects to per-topic mailboxes whose name
    is ``<broker_name>_topic_<topic>``.
    """

    def __init__(self, name: str = "mqtt_broker") -> None:
        self.broker_name = name
        self._subscriptions: Dict[str, List[str]] = {}
        self._messages_published = 0
        self._messages_delivered = 0
        self._running = True

    # -- SimGrid actor entry point ------------------------------------------ #

    def __call__(self) -> None:
        host = simgrid.this_actor.get_host()
        simgrid.this_actor.info(
            f"MQTT Broker '{self.broker_name}' started on host '{host.name}'"
        )
        ctrl_mbox = simgrid.Mailbox.by_name(self._ctrl_mbox_name)

        while self._running:
            try:
                comm = ctrl_mbox.get_async()
                comm.wait_for(10.0)          # 10 s idle → exit
                ctrl = comm.get_payload()
            except (simgrid.TimeoutException, simgrid.NetworkFailureException):
                simgrid.this_actor.info("Broker idle timeout – shutting down")
                break
            if not isinstance(ctrl, _ControlMessage):
                simgrid.this_actor.warning(
                    f"Broker received unexpected object type: {type(ctrl)}"
                )
                continue

            if ctrl.ctrl_type == _ControlType.SUBSCRIBE:
                self._handle_subscribe(ctrl.topic, ctrl.subscriber)
            elif ctrl.ctrl_type == _ControlType.UNSUBSCRIBE:
                self._handle_unsubscribe(ctrl.topic, ctrl.subscriber)
            elif ctrl.ctrl_type == _ControlType.PUBLISH:
                self._handle_publish(ctrl.message)
            elif ctrl.ctrl_type == _ControlType.SHUTDOWN:
                simgrid.this_actor.info("Broker shutdown requested")
                self._running = False

        self._print_stats()
        simgrid.this_actor.info(f"MQTT Broker '{self.broker_name}' terminated")

    # -- Internal helpers --------------------------------------------------- #

    def _handle_subscribe(self, topic: str, subscriber: str) -> None:
        subs = self._subscriptions.setdefault(topic, [])
        if subscriber not in subs:
            subs.append(subscriber)
            simgrid.this_actor.info(
                f"Subscriber '{subscriber}' subscribed to topic '{topic}' "
                f"({len(subs)} subscribers)"
            )

    def _handle_unsubscribe(self, topic: str, subscriber: str) -> None:
        if topic in self._subscriptions:
            try:
                self._subscriptions[topic].remove(subscriber)
            except ValueError:
                pass
            if not self._subscriptions[topic]:
                del self._subscriptions[topic]

    def _handle_publish(self, msg: MQTTMessage) -> None:
        self._messages_published += 1
        simgrid.this_actor.info(
            f"Publishing to topic '{msg.topic}' ({msg.size} bytes, from '{msg.publisher}')"
        )
        subs = self._subscriptions.get(msg.topic, [])
        if not subs:
            simgrid.this_actor.info(f"No subscribers for topic '{msg.topic}'")
            return
        for sub_mbox_name in subs:
            dest = simgrid.Mailbox.by_name(sub_mbox_name)
            dest.put(msg, msg.size)
            self._messages_delivered += 1

    def _print_stats(self) -> None:
        simgrid.this_actor.info(
            f"Broker stats – published: {self._messages_published}, "
            f"delivered: {self._messages_delivered}"
        )

    # -- Static helpers ------------------------------------------------------ #

    @property
    def _ctrl_mbox_name(self) -> str:
        return MQTTBroker.get_broker_mailbox(self.broker_name)

    @staticmethod
    def get_broker_mailbox(broker_name: str = "mqtt_broker") -> str:
        """Return the control mailbox name for a given broker."""
        return f"{broker_name}_ctrl"

    @staticmethod
    def get_topic_mailbox(broker_name: str, topic: str) -> str:
        """Return the delivery mailbox name for a given broker+topic pair."""
        safe_topic = topic.replace("/", "_").replace("#", "all").replace("+", "any")
        return f"{broker_name}_topic_{safe_topic}"


# --------------------------------------------------------------------------- #
# Convenience launcher                                                         #
# --------------------------------------------------------------------------- #

def start_broker(host: "simgrid.Host", broker_name: str = "mqtt_broker") -> None:
    """
    Convenience function: launch an :class:`MQTTBroker` actor on *host*.

    Call this once before any publishers or subscribers are created::

        start_broker(broker_host, "mqtt_broker")
    """
    simgrid.Actor.create(f"mqtt_broker_{broker_name}", host, MQTTBroker(broker_name))
